### PR TITLE
add serialized simulation to diagnostics files

### PR DESCRIPTION
**Pull request by PhilipDeegan**
how to use

```
import pyphare.pharein as ph
from pyphare.pharesee.run import Run

b = Run(".").GetB(0)
sim = ph.simulation.deserialize(b.data_files["py_attrs"].attrs["serialized_simulation"])
print(sim.electrons.closure.Te)
```

this next form is provided for convenience

```
import pyphare.pharein as ph
from pyphare.pharesee.run import Run

b = Run(".").GetB(0)
print(b.sim.electrons.closure.Te)
```

## Summary by CodeRabbit

- **New Features**
	- Introduce functionality to store and utilize serialized simulation data for diagnostics and restarts.

- **Refactor**
	- Minor code reorganization and formatting improvements in diagnostics handling.
	- Removed an unnecessary empty line for code cleanliness.

- **Tests**
	- Enhanced diagnostics tests to include checks for serialized simulation data integrity.

### CODE REVIEW

File: pyphare/pyphare/pharein/__init__.py [hunk 1]
```diff
@@ -251,9 +251,14 @@ def as_paths(rb):
     addInitFunction(maginit_path + "y_component", fn_wrapper(modelDict["by"]))
     addInitFunction(maginit_path + "z_component", fn_wrapper(modelDict["bz"]))
 
+    serialized_sim = serialize_sim(simulation)
+
     #### adding diagnostics
+
     diag_path = "simulation/diagnostics/"
     for diag in list(simulation.diagnostics.values()):
+        diag.attributes["serialized_simulation"] = serialized_sim
```
Comment (PhilipDeegan) [inline pyphare/pyphare/pharein/__init__.py:260]: this doesn't have to be like this exactly, but this does work, and is convenient.

File: pyphare/pyphare/pharesee/hierarchy.py [hunk 2]
```diff
@@ -608,6 +609,21 @@ def __init__(
...
+
+        if "py_attrs" not in self.data_files:
+            raise ValueError("Simulation is not available for deserialization")
+
+        from ..pharein.simulation import deserialize
+
+        self._sim = deserialize(
+            self.data_files["py_attrs"].attrs["serialized_simulation"]
+        )
+        return self._sim
```
[the remark this answers is not in the document]
Comment (PhilipDeegan) [reply pyphare/pyphare/pharesee/hierarchy.py:625]: good suggestion 🐰

**Review by rochSmets — APPROVED** [review]

**Comment by rochSmets** [comment]
Fine to me... But @PhilipDeegan you will have to explain me where the detail of what is serialized is written... I couldn't find where

**Comment by nicolasaunai** [comment]
@rochSmets it's there : https://github.com/PHAREHUB/PHARE/blob/master/pyphare/pyphare/pharein/simulation.py#L921-L931

The whole simulation object is encoded and written as a string by the package `dill`, which on demand can thus re-read it.

**Pull request merged**